Make documentation build work, and improve documentation for Win32.

al45tair opened the pull request:
Conditionalise the implementation for the Win32 executors so that the documentation can be built from Darwin but will still include the Win32 native executors.

Also improve the doc comments for those executors.

Additionally, add the initial implementation of `PlatformExecutorFactory`.

al45tair committed:
Make documentation build work, and improve documentation for Win32.

Conditionalise the implementation for the Win32 executors so that the
documentation can be built from Darwin but will still include the
Win32 native executors.

Also improve the doc comments for those executors.

al45tair committed:
Formatting fixes

FranzBusch commented on Package.swift:
```diff
@@ -3,6 +3,9 @@ import PackageDescription
   name: "SwiftPlatformExecutors",
+  platforms: [
+    .macOS(.v26)
+  ],
```
Can we avoid adding any platform requirements here and instead use availability annotations? Platform requirements will make it next to impossible for other packages that have lower requirements to adopt it.

al45tair replied:
Good thought.

FranzBusch commented on Sources/PlatformExecutors/PlatformExecutorFactory.swift:
```diff
@@ -0,0 +1,34 @@
+    // PThreadMainExecutor()
+  }
+  public static var defaultExecutor: any TaskExecutor {
+    // FIXME: The default pool size should be sensible for the hardware
```
Yes that's on my list to get the correct core count also accounting for container environments

FranzBusch commented on Sources/PlatformExecutors/PlatformExecutorFactory.swift:
```diff
@@ -0,0 +1,34 @@
+  public static var mainExecutor: any MainExecutor { Win32EventLoopExecutor(isMainExecutor: true) }
+  public static var defaultExecutor: any TaskExecutor { Win32ThreadPoolExecutor() }
+}
+#elseif os(Linux) || os(FreeBSD) || canImport(Darwin)
```
For Darwin we need to pick a different implementation I assume

al45tair replied:
Ideally, though the Pthread code will work, no?

FranzBusch replied:
Yeah it will but I'm not sure we should pick that as the platform's default for Darwin

al45tair replied:
Maybe the thing to do here is to have the `#else` case use `_Concurrency.DefaultExecutorFactory`. That would make some kind of sense, I suppose.

FranzBusch replied:
Yeah I think that makes sense

FranzBusch commented on Sources/PlatformExecutors/Win32Executor/Win32NativeExecutors.swift:
```diff
@@ -264,32 +370,37 @@ public final class Win32EventLoopExecutor: SerialExecutor, RunLoopExecutor, @unc
 
   deinit {
+    #if canImport(WinSDK)
     unsafe CloseHandle(hEvent)
```
Oh I just noticed this. For the platform executors that's probably fine but we should avoid doing any resource clean ups in deinit's. I was already thinking about this on the Linux side but we most likely need to offer those executors in a `with-style` like API. Can be done in a follow up though.

al45tair replied:
My thinking on this is that for the specific case of executors, it's almost always going to be fine to do cleanup in `deinit`, the reason being that you don't often create and destroy them so the chances of ending up in a situation where you've got a resource problem as a result are small.

FranzBusch replied:
From experience I don't think that's true sadly. We do have server users that are spinning up event loops and destroying them. We can do this in a follow up though.

FranzBusch commented on Sources/PlatformExecutors/Win32Executor/Win32NativeExecutors.swift:
```diff
@@ -12,6 +12,86 @@
 
 #if canImport(WinSDK)
 import WinSDK
+#else
```
Should we gate this behind some compiler condition that we check for docs. Otherwise we will commit those to our public API

al45tair replied:
We should. I haven't done that yet.

al45tair replied:
I've been thinking about this, and most of these are already marked as not visible in the documentation. I wanted `MSG`  and `PTP_POOL` to be visible so that we can link to Microsoft's documentation though.

al45tair replied:
On second thought, no, that's not ideal because it means we'll export `MSG`, `PTP_POOL` et al as types on Linux and Darwin. We should have a define.

al45tair committed:
Add a default `PlatformExecutorFactory`.

If it isn't a platform we recognise, make the default `PlatformExecutorFactory`
be the one from `_Concurrency` itself.

al45tair committed:
More documentation tweaks

al45tair committed:
Increase minimum delay tolerance slightly.

We *just* failed by 0.1ms, so bump this up by another millisecond.

al45tair committed:
Only include Win32 things when either on Windows *or* building docs.

We want the Win32 stuff to show up when building documentation.
